packages/block-library/src/legacy-widget/edit/handler.js

```diff
@@ -45,78 +59,135 @@ class LegacyWidgetEditHandler extends Component {
+					// we don't want to unmount the component.
+					// Otherwise when we went back to edit we wound need to trigger
+					// all widgets events again and some scripts may not deal well with this.
+					style={ {
+						display: this.props.isVisible ? 'block' : 'none',
 					} }
-					onInstanceChange={ this.onInstanceChange }
-					widgetNumber={ instanceId * -1 }
-					id={ id }
-					idBase={ idBase }
-					form={ form }
-				/>
-			</div>
+				>
+
+					<LegacyWidgetEditDomManager
+						isReferenceWidget={ !! widgetId }
+						ref={ ( ref ) => {
+							this.widgetEditDomManagerRef = ref;
+						} }
+						onInstanceChange={ this.onInstanceChange }
+						widgetNumber={ widgetNumber ? widgetNumber : instanceId * -1 }
+						id={ widgetId }
+						idBase={ idBase }
+						form={ form }
+						widgetId={ widgetId }
+					/>
+				</div>
+			</>
 		);
 	}
 
+	trySetNonce() {
+		const element = document.getElementById( '_wpnonce_widgets' );
+		if ( element && element.value ) {
+			this.widgetNonce = element.value;
+		}
+	}
+
 	onInstanceChange( instanceChanges ) {
+		const { widgetId } = this.props;
+		if ( widgetId ) {
+			// For reference widgets there is no need to query an endpoint,
+			// the widget is already saved with ajax.
+			this.props.onInstanceChange( instanceChanges, true );
+			return;
+		}
 		this.requestWidgetUpdater( instanceChanges, ( response ) => {
 			this.instanceUpdating = response.instance;
-			this.props.onInstanceChange( response.instance );
+			this.props.onInstanceChange( response.instance, !! response.form );
 		} );
 	}
 
 	requestWidgetUpdater( instanceChanges, callback ) {
-		const { identifier, instanceId, instance } = this.props;
-		if ( ! identifier ) {
+		const { widgetId, idBase, instance, widgetClass } = this.props;
+		const { isStillMounted } = this;
+		if ( ! widgetId && ! widgetClass ) {
 			return;
 		}
 
-		apiFetch( {
-			path: `/wp/v2/widgets/${ identifier }/`,
-			data: {
-				identifier,
-				instance,
-				// use negative ids to make sure the id does not exist on the database.
-				id_to_use: instanceId * -1,
-				instance_changes: instanceChanges,
-			},
-			method: 'POST',
-		} ).then(
-			( response ) => {
-				if ( this.isStillMounted ) {
-					this.setState( {
-						form: response.form,
-						idBase: response.id_base,
-						id: response.id,
-					} );
+		// If we are in the presence of a reference widget, do a save ajax request
+		// with empty changes so we retrieve the widget edit form.
+		if ( widgetId ) {
+			const httpRequest = new XMLHttpRequest();
+			const formData = new FormData();
+			formData.append( 'action', 'save-widget' );
+			formData.append( 'id_base', idBase );
+			formData.append( 'widget-id', widgetId );
+			formData.append( 'widget-width', '250' );
+			formData.append( 'widget-height', '200' );
+			formData.append( 'savewidgets', this.widgetNonce );
+			httpRequest.open( 'POST', window.ajaxurl );
+			const self = this;
```

Comment:
Let's use a `=>` function instead of a `function()` function below so that `const self = this;` isn't necessary.